File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/tests/iteration/grammars_test.py
import pytest
from gbnf import GBNF


def describe_grammars():
    def _load_cases():
        import json
        from pathlib import Path

        grammars_dir = Path(__file__).parent / "grammars"
        cases = []
        for grammar_path in sorted(grammars_dir.glob("*.gbnf")):
            json_path = grammar_path.with_suffix(".json")
            grammar = grammar_path.read_text()
            for test_case in json.loads(json_path.read_text()):
                cases.append((grammar_path.stem, test_case, grammar))
        if not cases:
            raise Exception("No grammar test cases found")
        return cases

    @pytest.mark.parametrize(("key", "test_case", "grammar"), _load_cases())
    def test_it_parses_a_known_valid_grammar(key, test_case, grammar):
        state = GBNF(grammar)
        for char in test_case:
            state = state.add(char)

File: packages/gbnf-experiment/data/20260909T022216Z_19e0fe30/ported_implementation/tests/iteration/grammars.test.ts
import { readdirSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, expect, it } from 'vitest';

import { GBNF } from '../../src/index.js';

const grammarsDir = join(dirname(fileURLToPath(import.meta.url)), 'grammars');

const loadCases = (): [string, string, string][] => {
  const cases: [string, string, string][] = [];
  const grammarFiles = readdirSync(grammarsDir)
    .filter((file) => file.endsWith('.gbnf'))
    .sort();
  for (const file of grammarFiles) {
    const key = file.replace(/\.gbnf$/, '');
    const grammar = readFileSync(join(grammarsDir, file), 'utf8');
    const testCases = JSON.parse(
      readFileSync(join(grammarsDir, `${key}.json`), 'utf8'),
    ) as string[];
    for (const testCase of testCases) {
      cases.push([key, testCase, grammar]);
    }
  }
  if (cases.length === 0) {
    throw new Error('No grammar test cases found');
  }
  return cases;
};

describe('grammars', () => {
  it.each(loadCases().map((c, idx) => [`${c[0]} #${idx}`, c]))(
    'it parses a known valid grammar: %s',
    (_name, values) => {
      const [, testCase, grammar] = values as [string, string, string];
      let state = GBNF(grammar);
      for (const char of testCase) {
        state = state.add(char);
      }
      expect(state).toBeDefined();
    },
  );
});
